chiselFrontend/src/main/scala/chisel3/core/RawModule.scala

```diff
@@ -124,7 +124,11 @@ abstract class MultiIOModule(implicit moduleCompileOptions: CompileOptions)
     extends RawModule {
   // Implicit clock and reset pins
   val clock: Clock = IO(Input(Clock()))
-  val reset: Reset = IO(Input(Bool()))
+  val reset: Reset = {
+    // Default if no outer reset is synchronous reset
+    val gen = Builder.outerReset.map(_.cloneType).getOrElse(Bool())
```

Comment:
The old way of assigning clock and reset will work with async reset, but only if the implicit reset is AsyncReset. I grant that this example could trip someone up, and it's probably something people will try since you'll implicitly have a Bool as reset when you have a Module as your top module. I don't really know what else we can do though without sacrificing what little type safety we're able to have 🤷‍♂️

Reply:
I added a test that illustrates this error

Reply:
I'm still a bit uncomfortable with the structure of this, where a Module's IO signature (class of reset field) is dependent implicitly on the enclosing Module's reset type. Also, is this getting you any more type safety? As far as the type checker is concerned `val reset` is of type Reset, and only runtime type information can determine whether that's a AsyncReset or (synchronous) Bool.

Also, the potential for reassignment of reset makes me feel that determining whether a reset is sync or async is a concern that's downstream of Chisel, since (generally speaking) that would need access to the hardware graph that's not available in Chisel. As implemented this handles a special case (though probably the most common one) where reset is not reassigned.

And if sync and async reset versions of a Module are used simultaneously, this would generate two different Modules even in HiFIRRTL? Should that be the case, and is that intuitive for debugging?

Reply:
Including Chisel's "type system" as a form of type safety, this proposal is no less "type safe" than most things in Chisel (see connections, etc.), while still providing nice flexibility. Would you rather than we have to have both a SyncResetQueue and an AsyncResetQueue and you have to statically pick the right one? Similarly to how Chisel doesn't make you worry about clock and reset if you're in the common case of only having one of both, in my opinion it's really nice to let the context within which a Module is used specify the type of the implicit reset (as long as it supports both).

> Also, the potential for reassignment of reset makes me feel that determining whether a reset is sync or async is a concern that's downstream of Chisel

I'm not sure I understand what you're proposing by determining sync vs. async reset downstream of Chisel. Async reset and synchronous reset are not functionally equivalent. Designers need the ability to specify the type of the reset, but should also be empowered to write reset-agnostic *generators*.

> And if sync and async reset versions of a Module are used simultaneously, this would generate two different Modules even in HiFIRRTL? Should that be the case, and is that intuitive for debugging?

Similarly to the above, a module using sync reset and a module using async reset are functionally different and should be treated as such in FIRRTL.

My proposal in this PR is to make it possible to express a Module such that it works with either (as in the case of the Queue instantiation example), or may only work with synchronous reset (as caught by the error checking for reset values being non-literals, and leaving catching safe reset treatment for a FIRRTL transformation akin to checking for safe clock crossings). The idea is that you can use the Scala type system if you want, by making your resets be of type `Bool`, `AsyncReset`, or `Reset` (for both). For convenience and backwards compatibility reasons, I thought it prudent to make `Module` work with either. Many modules are written to work with both so this is fairly convenient. And for those that won't work in this way, it will be caught by the checks mentioned above.

One thing that I *do* think is currently missing from this PR is an easy and explicit way to constraint the type of the implicit reset. Currently, the best way to do so would be:
```scala
require(Module.reset.isInstanceOf[AsyncReset])
// or
require(Module.reset.isInstanceOf[Bool])
```
I would appreciate suggestions for improvement if anyone has ideas! Some other possibilities:
```scala
require(Module.resetIsAsync)
require(Module.resetIsSync)
// or
require(Module.resetIsBool)
```
Perhaps a mixin for `Module` like:
```scala
class MyModule extends Module with ImplicitAsyncReset { ... }
```
I generally like the idea of a mixin but it doesn't work utility functions and whatnot that may require a specific type of reset.

Reply:
I think the type safety of Reset (general reset, that could be either sync or async), AsyncReset and Bool (synchronous reset) makes sense. But what about making the Reset type instantiable so that you could create a Module with an IO of type Reset that indicates that the Module could be generated as either sync or async reset? This seems to better encode user intent, and could also allow better linking of Modules in FIRRTL since whether the Module is sync or async is encoded in FIRRTL and you can delay the actual reset resolution to link time? And inter-module linking is really kind of in the FIRRTL domain (with Bindings being an exception, though the motivation is largely usability instead of separation of responsibilities).

I also do like the mixin approach, perhaps the mixin could hook into the Module reset IO generation logic and force it to a subtype of Reset? It would be very tight coupling between ImplicitAsyncReset/ImplicitSyncReset and Module, though the exposed API is nice.

Reply:
See below for plan of record